gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx

```diff
@@ -0,0 +1,271 @@
+    TypeVar "_T@list" is invariant
+      Tuple entry 2 is incorrect type
+        Tuple size mismatch; expected 3 but received indeterminate number (reportGeneralTypeIssues)
+  ```
+  This is really valuable information. It's worth considering Pyright as an option to get a second opinion!
+- It looks like Pyright's name for `Any` is `Unknown`. I think that does a better job of emphasising that `Unknown` won't be type checked. I'd certainly be more reluctant to type `x: Unknown` versus `x: Any`!
+- Pyright is the machinery behind [Pylance](https://github.com/microsoft/pylance-release), which drives [VS Code's Python extension](https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance). That alone probably makes it worthy of more eyes.
+- Seemed like it was the best-placed alternative typechecker to challenge mypy (the de-facto standard).
+
+
+#### [Pytype](https://github.com/google/pytype) (Google)
+- Google internal? Seems to be maintained by one person semiregularly by "syncing" from Google.
+- Apparently contains a script [`merge-pyi`](https://github.com/google/pytype/tree/master/pytype/tools/merge_pyi) to annotate a source file given a stub file.
+- No support for TypedDict: as soon as it saw one in Sydent, it stopped all analysis.
+- No Python 3.10 support (according to the README anyway).
+- I think it might use a different kind of typing semantics; its [typing FAQ](https://google.github.io/pytype/typing_faq.html) speaks of "descriptive typing" and a more lenient approach.
+
+#### [PyCharm](https://www.jetbrains.com/pycharm/)
+
+PyCharm has its own means to typechecking code as you write it. It's definitely caught bugs before, and having the instant feedback as you type is really nice! I have seen it struggle with `zope.interface` and some uses of `Generic`s though.
+
+
+### Runtime uses of annotations
+
+When annotations [were first introduced](https://www.python.org/dev/peps/pep-3107/), they were a generic means to associate Python objects with parts of a program. (It was only later that the community agreed that we really want to use them to annotate _types_). These annotations are available at runtime in the `__annotations__` attribute. There's also a helper function in `typing` which will help resolve forward references.
+
+```python
+>>> from typing import get_type_hints
+>>> def foo(x: int) -> str: pass
+...
+>>> get_type_hints(foo)
+{'x': <class 'int'>, 'return': <class 'str'>}
+```
+
+Programs and libraries are free to use these annotations at runtime as they see fit. The most well-known examples are probably [`dataclasses`](), [`attrs`](https://www.attrs.org/en/stable/) with [`auto_attribs=True`](https://www.attrs.org/en/stable/types.html) and [`Pydantic`](https://pydantic-docs.helpmanual.io/). I'd be interested to learn if anyone else is consuming annotations at runtime!
```

Comment:
Did you want to add a url to the `dataclasses` link?

Reply:
I did, thank you!